resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesExternalShuffleService.scala

```diff
@@ -137,42 +134,54 @@ private[spark] class KubernetesShuffleBlockHandler (
  */
 private[spark] class KubernetesExternalShuffleService(
     conf: SparkConf,
-    securityManager: SecurityManager,
-    kubernetesClientProvider: DriverPodKubernetesClientProvider)
+    securityManager: SecurityManager)
   extends ExternalShuffleService(conf, securityManager) {
 
   private var shuffleBlockHandlers: mutable.Buffer[KubernetesShuffleBlockHandler] = _
   protected override def newShuffleBlockHandler(
       tConf: TransportConf): ExternalShuffleBlockHandler = {
-    val newBlockHandler = new KubernetesShuffleBlockHandler(tConf, kubernetesClientProvider)
-    newBlockHandler.start()
-
-    // TODO: figure out a better way of doing this.
-    // This is necessary because the constructor is not called
-    // when this class is initialized through ExternalShuffleService.
-    if (shuffleBlockHandlers == null) {
+    val kubernetesClient = SparkKubernetesClientFactory.createKubernetesClient(
+        conf.get(KUBERNETES_SHUFFLE_APISERVER_URI),
+        None,
+        APISERVER_AUTH_SHUFFLE_SERVICE_CONF_PREFIX,
+        conf,
+        Some(new File(Config.KUBERNETES_SERVICE_ACCOUNT_TOKEN_PATH))
+            .filter( _ => conf.get(KUBERNETES_SHUFFLE_USE_SERVICE_ACCOUNT_CREDENTIALS)),
+        Some(new File(Config.KUBERNETES_SERVICE_ACCOUNT_CA_CRT_PATH))
+            .filter( _ => conf.get(KUBERNETES_SHUFFLE_USE_SERVICE_ACCOUNT_CREDENTIALS)))
+    val newBlockHandler = new KubernetesShuffleBlockHandler(tConf, kubernetesClient)
+    try {
+      newBlockHandler.start()
+      // TODO: figure out a better way of doing this.
+      // This is necessary because the constructor is not called
+      // when this class is initialized through ExternalShuffleService.
```

Comment:
I think the comment isn't clear there. The issue I think, was that the `ExternalShuffleService` is extended by the `KubernetesExternalShuffleService`, but the `ExternalShuffleService` calls an overridden method in its own constructor. So, at that point, the constructor for `KubernetesExternalShuffleService` has not yet been called.